### PR TITLE
COR-581 individual form

### PR DESCRIPTION
**Pull request by ludydoo**
- Adding the migration to create the individual form
- Use uuids for ids instead of xid
- Add the visual form
- Removes unused constants for old fields that don't exist anymore
- Changes the mock data generator to generate data that corresponds to the new form

### CODE REVIEW

File: internal/db/country.go [hunk 1]
```diff
@@ -88,7 +88,7 @@ func (c countryRepo) updateCountry(ctx context.Context, country *api.Country) (*
 func (c countryRepo) createCountry(ctx context.Context, country *api.Country) (*api.Country, error) {
 	l := c.logger(ctx)
 	l.Debug("creating new country")
-	country.ID = xid.New().String()
+	country.ID = uuid.New().String()
```
Comment (internetti) [inline internal/db/country.go:91]: no issue here, but why the change?

File: internal/db/migrations/postgres/012_new_individual.up.sql [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+DROP TABLE IF EXISTS individual_registrations;
+CREATE TABLE individual_registrations
+(
+    address                           varchar(512)             NOT NULL,
+    birth_date                        date                     NULL,
+    cognitive_disability_level        varchar(32)              NOT NULL,
+    collection_administrative_area_1  varchar(128)             NOT NULL,
+    collection_administrative_area_2  varchar(128)             NOT NULL,
+    collection_administrative_area_3  varchar(128)             NOT NULL,
+    collection_agent_name             varchar(64)              NOT NULL,
```
Comment (internetti) [inline internal/db/migrations/postgres/012_new_individual.up.sql:28]: potentially a bit short?

File: internal/db/migrations/postgres/012_new_individual.up.sql [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    has_consented_to_referral         boolean                  NOT NULL,
+    has_hearing_disability            boolean                  NOT NULL,
+    has_mobility_disability           boolean                  NOT NULL,
+    has_selfcare_disability           boolean                  NOT NULL,
+    has_vision_disability             boolean                  NOT NULL,
+    hearing_disability_level          varchar(32)              NOT NULL,
+    household_id                      varchar(64)              NOT NULL,
+    id                                uuid                     NOT NULL,
+    identification_type_1             varchar(64)              NOT NULL,
+    identification_type_explanation_1 varchar(255)             NOT NULL,
```
Comment (internetti) [inline internal/db/migrations/postgres/012_new_individual.up.sql:52]: I could imagine that these need to be longer?

File: internal/db/migrations/postgres/012_new_individual.up.sql [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+/*
+this procedure is used to prevent updating soft-deleted
+records.
+
+this procedure will be triggered before any update
+on the individuals table.
+
+Internally, it compares the old and the new record.
+If any of the values are being changed, other than the
+deleted_at column, it will throw an error.
```
Comment (internetti) [inline internal/db/migrations/postgres/012_new_individual.up.sql:95]: does that mean that the deleted_at column of a soft-deleted record can be edited? why would that happen?

File: internal/views/individual_form.go [hunk 5]
```diff
@@ -2,296 +2,836 @@ package views
...
+		f.buildPersonalInfoSection,
+		f.buildContactInfoSection,
+		f.buildProtectionSection,
+		f.buildDisabilitiesSection,
+		f.buildDataColletionSection,
+	); err != nil {
+		return err
+	}
+
+	builders := []builderFuncs{
```
Comment (internetti) [inline internal/views/individual_form.go:54]: even if it's programmatically not problematic that this has the same name as in the block from line 35, it makes it harder to understand

File: internal/views/individual_form.go [hunk 6]
```diff
@@ -2,296 +2,836 @@ package views
...
+		return nil, err
+	}
+	return f, nil
+}
 
-	isNew := i.ID == ""
+func (f *IndividualForm) build() error {
+	type builderFuncs func() error
+
+	runBuilders := func(builders ...builderFuncs) error {
```
Comment (internetti) [inline internal/views/individual_form.go:35]: could this be renamed runSectionBuilders?

File: internal/views/individual_form.go [hunk 7]
```diff
@@ -2,296 +2,836 @@ package views
...
+	if !f.isNew() {
+		idField := &forms.IDField{
+			Name:        "id",
+			DisplayName: "ID",
+			QRCodeURL:   fmt.Sprintf("/countries/%s/individuals/%s/qr", f.individual.CountryID, f.individual.ID),
+		}
+		if err := idField.SetValue(f.individual.ID); err != nil {
+			return err
+		}
+		f.personalInfoSection.Fields = append(f.personalInfoSection.Fields, idField)
```
Comment (internetti) [inline internal/views/individual_form.go:196]: these lines could be replaced by calling buildField right?

File: internal/views/individual_form.go [hunk 8]
```diff
@@ -2,296 +2,836 @@ package views
...
-		Value:       i.Address,
-	}
+		DisplayName: "Residence Address",
+	}, f.contactInfoSection, f.individual.Address)
+}
 
-	presentsProtectionConcernsField := &forms.CheckboxInputField{
+func (f *IndividualForm) buildPreferredMeansOfContact() error {
+	options := []forms.SelectInputFieldOption{
+		{Label: "WhatsApp", Value: "whatsapp"},
```
Comment (internetti) [inline internal/views/individual_form.go:430]: can we get Fix to confirm this list?

File: internal/views/individual_form.go [hunk 9]
```diff
@@ -2,296 +2,836 @@ package views
...
+	return buildField(&forms.SelectInputField{
+		Name:        "identificationContext",
+		DisplayName: "Identification Context",
+		Options: []forms.SelectInputFieldOption{
+			{Label: "", Value: ""},
+			{Label: "House Visit", Value: "houseVisit"},
+			{Label: "Field Activity", Value: "fieldActivity"},
+			{Label: "In-Office", Value: "inOffice"},
+			{Label: "Remote Channels", Value: "remoteChannels"},
+			{Label: "Other", Value: "other"},
```
Comment (internetti) [inline internal/views/individual_form.go:601]: This is missing 'referred' if I'm interpreting the excel sheet correctly

File: internal/views/individual_form.go [hunk 10]
```diff
@@ -2,296 +2,836 @@ package views
...
 
-	action := "/countries/" + i.CountryID + "/individuals/"
-	if len(i.ID) != 0 {
-		action += i.ID
-	} else {
-		action += "new"
+func getIdentificationTypeOptions() []forms.SelectInputFieldOption {
+	return []forms.SelectInputFieldOption{
+		{Label: "", Value: ""},
+		{Label: "Passport", Value: "passport"},
```
Comment (internetti) [inline internal/views/individual_form.go:671]: where did you get this list? is this confirmed?

File: internal/views/individual_form.go [hunk 11]
```diff
@@ -2,296 +2,836 @@ package views
...
-	return &IndividualForm{
-		Title: title,
-		Form:  &f,
+type displacementStatusCodec struct{}
+
+func (d *displacementStatusCodec) Encode(v interface{}) (string, error) {
+	switch v.(type) {
+	case api.DisplacementStatus:
+		switch v.(api.DisplacementStatus) {
+		case api.DisplacementStatusIDP:
```
Comment (internetti) [inline internal/views/individual_form.go:736]: this is missing the options 'stateless' and 'non-displaced'

**Review by internetti — COMMENTED** [review]
could you update the template file that users can download?

and can you add something to the readme how to run the migrations?

File: internal/views/individual_form.go (hunk 10 above)
Comment (ludydoo) [reply internal/views/individual_form.go:668]: It is not, but I just wanted to put some placeholders there until we have a final list

File: internal/views/individual_form.go (hunk 8 above)
Comment (ludydoo) [reply internal/views/individual_form.go:426]: Will ask

File: internal/db/country.go (hunk 1 above)
Comment (ludydoo) [reply internal/db/country.go:91]: the `xid` is useful when we want to have consecutive IDs (sortable by generation time). But we now have the "created_at" field that can do that.  And UUIDs are more widely used...

File: internal/views/individual_form.go [hunk 12]
```diff
@@ -2,296 +2,841 @@ package views
...
+	}, f.personalInfoSection, f.individual.PrefersToRemainAnonymous)
+}
 
-	if isNew {
+func (f *IndividualForm) buildGender() error {
+	genderOptions := getGenderOptions()
+	if f.isNew() {
 		genderOptions = append([]forms.SelectInputFieldOption{
 			{
 				Value: "",
```
Comment (internetti) [inline internal/views/individual_form.go:228]: when going to the page to create a new individual, this causes an error, 'unknown gender'

Comment (ludydoo) [reply internal/views/individual_form.go:228]: Made a mistake.. fixed the issue in COR-568 :/

File: internal/views/individual_form.go (hunk 8 above)
Comment (internetti) [reply internal/views/individual_form.go:426]: assuming you don't wanna wait to merge this until we get an answer, you wanna create a new ticket, so we don't forget to update these?

File: internal/db/migrations/postgres/012_new_individual.up.sql (hunk 4 above)
Comment (internetti) [reply internal/db/migrations/postgres/012_new_individual.up.sql:95]: any comment on this?

File: internal/views/individual_form.go (hunk 5 above)
Comment (internetti) [reply internal/views/individual_form.go:54]: opinions?

File: internal/views/individual_form.go (hunk 8 above)
Comment (ludydoo) [reply internal/views/individual_form.go:426]: Added ticket COR-645 to track this

**Comment by ludydoo** [comment]
> could you update the template file that users can download?

done 

> and can you add something to the readme how to run the migrations?

the migrations are run automatically when starting the server

**Review by internetti — APPROVED** [review]
